doc/unbound-checkconf.8
.TH "unbound-checkconf" "8" "@date@" "NLnet Labs" "unbound @version@"
.\"
.\" unbound-checkconf.8 -- unbound configuration checker manual
.\"
.\" Copyright (c) 2007, NLnet Labs. All rights reserved.
.\"
.\" See LICENSE for the license.
.\"
.\"
.SH "NAME"
.LP
unbound-checkconf
\- Check unbound configuration file for errors.
.SH "SYNOPSIS"
.B unbound-checkconf
.RB [ \-h ]
.IR cfgfile
.SH "DESCRIPTION"
.B Unbound-checkconf
checks the configuration file for the
\fIunbound\fR(8)
DNS resolver for syntax and other errors. 
The config file syntax is described in 
\fIunbound.conf\fR(5).
.P
The available options are:
.TP
.B \-h
Show the version and commandline option help.
.TP
.I cfgfile
The config file to read with settings for unbound. It is checked.
If omitted, the config file at the default location is checked.
.SH "EXIT CODE"
The unbound-checkconf program exits with status code 1 on error, 
0 for a correct config file.
.SH "FILES"
.TP
.I /usr/local/etc/unbound/unbound.conf
unbound configuration file.
.SH "SEE ALSO"
\fIunbound.conf\fR(5), 
\fIunbound\fR(8).


doc/Changelog
25 April 2008: Wouter
	- DESTDIR is honored by the Makefile for rpms.
	- contrib files unbound.spec and unbound.init, builds working RPM
	  on FC7 Linux, a chrooted caching resolver, and libunbound.
	- iana ports update.

24 April 2008: Wouter
	- chroot checks improved. working directory relative to chroot.
	  checks if config file path is inside chroot. Documentation on it.
	- nicer example.conf text.
	- created 0.11 tag.

23 April 2008: Wouter
	- parseunbound.pl contrib update from Kai Storbeck for threads.
	- iana ports update

22 April 2008: Wouter
	- ignore SIGPIPE.
	- unit test for SIGPIPE ignore.

21 April 2008: Wouter
	- FEATURES document.
	- fixup reread of config file if it was given as a full path
	  and chroot was used.

16 April 2008: Wouter
	- requirements doc, updated clean query returns.
	- parseunbound.pl update from Kai Storbeck.
	- sunos4 porting changes.

15 April 2008: Wouter
	- fixup default rc.d pidfile location to /usr/local/etc.
	- iana ports updated.
	- copyright updated in ldns-testpkts to keep same as in ldns.
	- fixup checkconf chroot tests a bit more, chdir must be inside
	  chroot dir.
	- documented 'gcc: unrecognized -KPIC option' errors on Solaris.
	- example.conf values changed to /usr/local/etc/unbound
	- DSA test work.
	- DSA signatures: unbound is compatible with both encodings found.
	  It will detect and convert when necessary.

14 April 2008: Wouter
	- got update for parseunbound.pl statistics script from Kai Storbeck.
	- tpkg tests for udp wait list.
	- documented 0x20 status.
	- fixup chroot and checkconf, it is much smarter now.
	- fixup DSA EVP signature decoding. Solution that Jelte found copied.
	- and check first sig byte for the encoding type.

11 April 2008: Wouter
	- random port selection out of the configged ports.
	- fixup threadsafety for libevent-1.4.3+ (event_base_get_method).
	- removed base_port.
	- created 256-port ephemeral space for the OS, 59802 available.
	- fixup consistency of port_if out array during heavy use.

10 April 2008: Wouter
	- --with-libevent works with latest libevent 1.4.99-trunk.
	- added log file statistics perl script to contrib.
	- automatic iana ports update from makefile. 60058 available.

9 April 2008: Wouter
	- configure can detect libev(from its build directory) when passed
	  --with-libevent=/home/wouter/libev-3.2
	  libev-3.2 is a little faster than libevent-1.4.3-stable (about 5%).
	- unused commpoints not listed in epoll list.
	- statistics-cumulative option so that the values are not reset.
	- config creates array of available ports, 61841 available,
	  it excludes <1024 and iana assigned numbers.
	  config statements to modify the available port numbers.

8 April 2008: Wouter
	- unbound tries to set the ulimit fds when started as server.
	  if that does not work, it will scale back its requirements.

27 March 2008: Wouter
	- documented /dev/random symlink from chrootdir as FAQ entry.

26 March 2008: Wouter
	- implemented AD bit signaling. If a query sets AD bit (but not DO)
	  then the AD bit is set in the reply if the answer validated.
	  Without including DNSSEC signatures. Useful if you have a trusted
	  path from the client to the resolver. Follows dnssec-updates draft.

25 March 2008: Wouter
	- implemented check that for NXDOMAIN and NOERROR answers a query
	  section must be present in the reply (by the scrubber). And it must
	  be equal to the question sent, at least lowercase folded.
	  Previously this feature happened because the cache code refused
	  to store such messages. However blocking by the scrubber makes 
	  sure nothing gets into the RRset cache. Also, this looks like a
	  timeout (instead of an allocation failure) and this retries are
	  done (which is useful in a spoofing situation).
	- RTT banding. Band size 400 msec, this makes band around zero (fast)
	  include unknown servers. This makes unbound explore unknown servers.

7 March 2008: Wouter
	- -C config feature for harvest program. 
	- harvest handles CNAMEs too.

5 March 2008: Wouter
	- patch from Hugo Koji Kobayashi for iterator logs spelling.

4 March 2008: Wouter
	- From report by Jinmei Tatuya, rfc2181 trust value for remainder
	  of a cname trust chain is lower; not full answer_AA. 
	- test for this fix.
	- default config file location is /usr/local/etc/unbound.
	  Thus prefix is used to determine the location. This is also the
	  chroot and pidfile default location.

3 March 2008: Wouter
	- Create 0.10 svn tag.
	- 0.11 version in trunk.
	- indentation nicer.

29 February 2008: Wouter
	- documentation update.
	- fixup port to Solaris of perf test tool.
	- updated ldns-tarball with decl-after-statement fixes.

28 February 2008: Wouter
	- fixed memory leaks in libunbound (during cancellation and wait).
	- libunbound returns the answer packet in full.
	- snprintf compat update.
	- harvest performs lookup.
	- ldns-tarball update with fix for ldns_dname_label.
	- installs to sbin by default.
	- install all manual pages (unbound-host and libunbound too).

27 February 2008: Wouter
	- option to use caps for id randomness.
	- config file option use-caps-for-id: yes
	- harvest debug tool

26 February 2008: Wouter
	- delay utility delays TCP as well. If the server that is forwarded 
	  to has a TCP error, the delay utility closes the connection.
	- delay does REUSE_ADDR, and can handle a server that closes its end.
	- answers use casing from query.

25 February 2008: Wouter
	- delay utility works. Gets decent thoughput too (>20000).

22 February 2008: Wouter
	- +2% for recursions, if identical queries (except for destination
	  and query ID) in the reply list, avoid re-encoding the answer.
	- removed TODO items for optimizations that do not show up in
	  profile reports.
	- default is now minievent - not libevent. As its faster and
	  not needed for regular installs, only for very large port ranges.
	- loop check different speedup pkt-dname-reading, 1% faster for
	  nocache-recursion check.

smallapp/unbound-checkconf.c
/*
 * checkconf/unbound-checkconf.c - config file checker for unbound.conf file.
 *
 * Copyright (c) 2007, NLnet Labs. All rights reserved.
 *
 * This software is open source.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * 
 * Neither the name of the NLNET LABS nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * \file
 *
 * The config checker checks for syntax and other errors in the unbound.conf
 * file, and can be used to check for errors before the server is started
 * or sigHUPped.
 * Exit status 1 means an error.
 */

#include "config.h"
#include "util/log.h"
#include "util/config_file.h"
#include "util/module.h"
#include "util/net_help.h"
#include "util/regional.h"
#include "iterator/iterator.h"
#include "validator/validator.h"
#include "services/localzone.h"
#include <pwd.h>
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif

/** Give checkconf usage, and exit (1). */
static void
usage()
{
	printf("Usage:	unbound-checkconf [file]\n");
	printf("	Checks unbound configuration file for errors.\n");
	printf("file	if omitted %s is used.\n", CONFIGFILE);
	printf("-h	show this usage help.\n");
	printf("Version %s\n", PACKAGE_VERSION);
	printf("BSD licensed, see LICENSE in source package for details.\n");
	printf("Report bugs to %s\n", PACKAGE_BUGREPORT);
	exit(1);
}

/** check if module works with config */
static void
check_mod(struct config_file* cfg, struct module_func_block* fb)
{
	struct module_env env;
	memset(&env, 0, sizeof(env));
	env.cfg = cfg;
	env.scratch = regional_create();
	env.scratch_buffer = ldns_buffer_new(BUFSIZ);
	if(!env.scratch || !env.scratch_buffer)
		fatal_exit("out of memory");
	if(!(*fb->init)(&env, 0)) {
		fatal_exit("bad config for %s module", fb->name);
	}
	(*fb->deinit)(&env, 0);
	ldns_buffer_free(env.scratch_buffer);
	regional_destroy(env.scratch);
}

/** check localzones */
static void
localzonechecks(struct config_file* cfg)
{
	struct local_zones* zs;
	if(!(zs = local_zones_create()))
		fatal_exit("out of memory");
	if(!local_zones_apply_cfg(zs, cfg))
		fatal_exit("failed local-zone, local-data configuration");
	local_zones_delete(zs);
}

/** emit warnings for IP in hosts */
static void
warn_hosts(const char* typ, struct config_stub* list)
{
	struct sockaddr_storage a;
	socklen_t alen;
	struct config_stub* s;
	struct config_strlist* h;
	for(s=list; s; s=s->next) {
		for(h=s->hosts; h; h=h->next) {
			if(extstrtoaddr(h->str, &a, &alen)) {
				fprintf(stderr, "unbound-checkconf: warning:"
				  " %s %s: \"%s\" is an IP%s address, "
				  "and when looked up as a host name "
				  "during use may not resolve.\n", 
				  s->name, typ, h->str,
				  addr_is_ip6(&a, alen)?"6":"4");
			}
		}
	}
}

/** check interface strings */
static void
interfacechecks(struct config_file* cfg)
{
	struct sockaddr_storage a;
	socklen_t alen;
	int i, j;
	for(i=0; i<cfg->num_ifs; i++) {
		if(!ipstrtoaddr(cfg->ifs[i], UNBOUND_DNS_PORT, &a, &alen)) {
			fatal_exit("cannot parse interface specified as '%s'",
				cfg->ifs[i]);
		}
		for(j=0; j<cfg->num_ifs; j++) {
			if(i!=j && strcmp(cfg->ifs[i], cfg->ifs[j])==0)
				fatal_exit("interface: %s present twice, "
					"cannot bind same ports twice.",
					cfg->ifs[i]);
		}
	}
	for(i=0; i<cfg->num_out_ifs; i++) {
		if(!ipstrtoaddr(cfg->out_ifs[i], UNBOUND_DNS_PORT, 
			&a, &alen)) {
			fatal_exit("cannot parse outgoing-interface "
				"specified as '%s'", cfg->out_ifs[i]);
		}
		for(j=0; j<cfg->num_out_ifs; j++) {
			if(i!=j && strcmp(cfg->out_ifs[i], cfg->out_ifs[j])==0)
				fatal_exit("outgoing-interface: %s present "
					"twice, cannot bind same ports twice.",
					cfg->out_ifs[i]);
		}
	}
}

/** check acl ips */
static void
aclchecks(struct config_file* cfg)
{
	int d;
	struct sockaddr_storage a;
	socklen_t alen;
	struct config_str2list* acl;
	for(acl=cfg->acls; acl; acl = acl->next) {
		if(!netblockstrtoaddr(acl->str, UNBOUND_DNS_PORT, &a, &alen, 
			&d)) {
			fatal_exit("cannot parse access control address %s %s",
				acl->str, acl->str2);
		}
	}
}

/** true if fname is a file */
static int
is_file(const char* fname) 
{
	struct stat buf;
	if(stat(fname, &buf) < 0) {
		if(errno==EACCES) {
			printf("warning: no search permission for one of the directories in path: %s\n", fname);
			return 1;
		}
		perror(fname);
		return 0;
	}
	if(S_ISDIR(buf.st_mode)) {
		printf("%s is not a file\n", fname);
		return 0;
	}
	return 1;
}

/** true if fname is a directory */
static int
is_dir(const char* fname) 
{
	struct stat buf;
	if(stat(fname, &buf) < 0) {
		if(errno==EACCES) {
			printf("warning: no search permission for one of the directories in path: %s\n", fname);
			return 1;
		}
		perror(fname);
		return 0;
	}
	if(!(S_ISDIR(buf.st_mode))) {
		printf("%s is not a directory\n", fname);
		return 0;
	}
	return 1;
}

/** convert a filename to full pathname in original filesys
 * @param fname: the path name to convert.
 * 	Must not be null or empty.
 * @param cfg: config struct for chroot and chdir (if set).
 * @param use_chdir: if false, only chroot is applied.
 * @return pointer to static buffer which is: [chroot][chdir]fname
 */
static char*
fname_after_chroot(const char* fname, struct config_file* cfg, int use_chdir)
{
	static char buf[1024];
	int slashit = 0;
	buf[0] = 0;
	if(cfg->chrootdir && cfg->chrootdir[0] && 
		strncmp(cfg->chrootdir, fname, strlen(cfg->chrootdir)) == 0) {
		/* already full pathname, return it */
		strncpy(buf, fname, sizeof(buf)-1);
		buf[sizeof(buf)-1] = 0;
		return buf;
	}
	/* chroot */
	if(cfg->chrootdir && cfg->chrootdir[0]) {
		/* start with chrootdir */
		strncpy(buf, cfg->chrootdir, sizeof(buf)-1);
		slashit = 1;
	}
	/* chdir */
	if(fname[0] == '/' || !use_chdir) {
		/* full path, no chdir */
	} else if(cfg->directory && cfg->directory[0]) {
		/* prepend chdir */
		if(slashit && cfg->directory[0] != '/')
			strncat(buf, "/", sizeof(buf)-1);
		if(strncmp(cfg->chrootdir, cfg->directory, 
			strlen(cfg->chrootdir)) == 0)
			strncat(buf, cfg->directory+strlen(cfg->chrootdir), 
				sizeof(buf)-1);
		else strncat(buf, cfg->directory, sizeof(buf)-1);
		slashit = 1;
	}
	/* fname */
	if(slashit && fname[0] != '/')
		strncat(buf, "/", sizeof(buf)-1);
	strncat(buf, fname, sizeof(buf)-1);
	buf[sizeof(buf)-1] = 0;
	return buf;
}

/** get base dir of a fname */
static char*
basedir(const char* fname, struct config_file* cfg)
{
	char* d = fname_after_chroot(fname, cfg, 1);
	char* rev = strrchr(d, '/');
	if(!rev) return NULL;
	if(d == rev) return NULL;
	rev[0] = 0;
	return d;
}

/** check file list, every file must be inside the chroot location */
static void
check_chroot_filelist(const char* desc, struct config_strlist* list,
	const char* chrootdir, struct config_file* cfg)
{
	struct config_strlist* p;
	char* old;
	for(p=list; p; p=p->next) {
		if(p->str && p->str[0]) {
			if(!is_file(fname_after_chroot(p->str, cfg, 1))) {
				fatal_exit("%s: \"%s\" does not exist in chrootdir %s", 
					desc, p->str, chrootdir);
			}
			old = p->str;
			/* put in a new full path for continued checking */
			p->str = strdup(fname_after_chroot(p->str, cfg, 1));
			free(old);
		}
	}
}

/** check configuration for errors */
static void
morechecks(struct config_file* cfg, char* fname)
{
	warn_hosts("stub-host", cfg->stubs);
	warn_hosts("forward-host", cfg->forwards);
	interfacechecks(cfg);
	aclchecks(cfg);

	if(cfg->verbosity < 0)
		fatal_exit("verbosity value < 0");
	if(cfg->num_threads < 0 || cfg->num_threads > 10000)
		fatal_exit("num_threads value weird");
	if(!cfg->do_ip4 && !cfg->do_ip6)
		fatal_exit("ip4 and ip6 are both disabled, pointless");
	if(!cfg->do_udp && !cfg->do_tcp)
		fatal_exit("udp and tcp are both disabled, pointless");

	if(cfg->chrootdir && cfg->chrootdir[0] && 
		cfg->chrootdir[strlen(cfg->chrootdir)-1] == '/')
		fatal_exit("chootdir %s has trailing slash '/' please remove.",
			cfg->chrootdir);
	if(cfg->chrootdir && cfg->chrootdir[0] && 
		!is_dir(cfg->chrootdir)) {
		fatal_exit("bad chroot directory");
	}
	if(cfg->chrootdir && cfg->chrootdir[0]) {
		char buf[10240];
		buf[0] = 0;
		if(fname[0] != '/') {
			if(getcwd(buf, sizeof(buf)) == NULL)
				fatal_exit("getcwd: %s", strerror(errno));
			strncat(buf, "/", sizeof(buf));
		}
		strncat(buf, fname, sizeof(buf));
		if(strncmp(buf, cfg->chrootdir, strlen(cfg->chrootdir)) != 0)
			fatal_exit("config file %s is not inside chroot %s",
				buf, cfg->chrootdir);
	}
	if(cfg->directory && cfg->directory[0] && !is_dir(
		fname_after_chroot(cfg->directory, cfg, 0))) {
		fatal_exit("bad chdir directory");
	}
	if( (cfg->chrootdir && cfg->chrootdir[0]) ||
	    (cfg->directory && cfg->directory[0])) {
		if(cfg->pidfile && cfg->pidfile[0] &&
		   basedir(cfg->pidfile, cfg) &&
		   !is_dir(basedir(cfg->pidfile, cfg))) {
			fatal_exit("pidfile directory does not exist");
		}
		if(cfg->logfile && cfg->logfile[0] &&
		   basedir(cfg->logfile, cfg) &&
		   !is_dir(basedir(cfg->logfile, cfg))) {
			fatal_exit("logfile directory does not exist");
		}
	}

	check_chroot_filelist("file with root-hints", 
		cfg->root_hints, cfg->chrootdir, cfg);
	check_chroot_filelist("trust-anchor-file", 
		cfg->trust_anchor_file_list, cfg->chrootdir, cfg);
	check_chroot_filelist("trusted-keys-file", 
		cfg->trusted_keys_file_list, cfg->chrootdir, cfg);
	/* remove chroot setting so that modules are not stripping pathnames*/
	free(cfg->chrootdir);
	cfg->chrootdir = NULL;
	
	if(strcmp(cfg->module_conf, "iterator") != 0 &&
		strcmp(cfg->module_conf, "validator iterator") != 0) {
		fatal_exit("module conf '%s' is not known to work",
			cfg->module_conf);
	}

	if(cfg->username && cfg->username[0]) {
		if(getpwnam(cfg->username) == NULL)
			fatal_exit("user '%s' does not exist.", cfg->username);
		endpwent();
	}

	localzonechecks(cfg);
}

/** check config file */
static void
checkconf(char* cfgfile)
{
	struct config_file* cfg = config_create();
	if(!cfg)
		fatal_exit("out of memory");
	if(!config_read(cfg, cfgfile)) {
		/* config_read prints messages to stderr */
		config_delete(cfg);
		exit(1);
	}
	morechecks(cfg, cfgfile);
	check_mod(cfg, iter_get_funcblock());
	check_mod(cfg, val_get_funcblock());
	config_delete(cfg);
	printf("unbound-checkconf: no errors in %s\n", cfgfile);
}

/** getopt global, in case header files fail to declare it. */
extern int optind;
/** getopt global, in case header files fail to declare it. */
extern char* optarg;

/** Main routine for checkconf */
int main(int argc, char* argv[])
{
	int c;
	char* f;
	log_ident_set("unbound-checkconf");
	log_init(NULL, 0, NULL);
	checklock_start();
	/* parse the options */
	while( (c=getopt(argc, argv, "h")) != -1) {
		switch(c) {
		case '?':
		case 'h':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;
	if(argc != 0 && argc != 1)
		usage();
	if(argc == 1)
		f = argv[0];
	else	f = CONFIGFILE;
	checkconf(f);
	checklock_stop();
	return 0;
}
